Implement missing Nether Blocks

pandaaaBE opened the pull request:
## Introduction
This plugin adds follow things:
- Crimson Nylium/Fungus, Waped Nylium/Fungus, Nether Sporuts
- Bone Meal Logic for Netherrack
- Adds support for Nylium to Mushrooms
- Adds support for Nylium to NetherRoots

Note: Bone Meal logic for Fungus is currently missing and is not planned in this pull request

Project Card: https://github.com/pmmp/PocketMine-MP/projects/14#card-84464878

(reopened due of unresolved conflicts of the old pr)

## Changes
### API changes
`VanillaBlocks::CRIMSON_NYLIUM(), VanillaBlocks::WARPED_NYLIUM(), VanillaBlocks::CRIMSON_FUNGUS(), VanillaBlocks::WARPED_FUNGUS(), VanillaBlocks::NETHER_SPROUTS()`

## Tests
<!--
PRs which have not been tested MUST be marked as draft.
-->
I tested this PR by doing the following (tick all that apply):
- [x] Writing PHPUnit tests (commit these in the `tests/phpunit` folder)
- [x] Playtesting using a Minecraft client (provide screenshots or a video)
![grafik](https://github.com/pmmp/PocketMine-MP/assets/135014092/8411dc60-a5af-45f9-9097-901ba6f91221)

pandaaaBE committed:
reupload all changed files from src/block

pandaaaBE committed:
reupload all changed files in src/block -> 2

pandaaaBE committed:
reupload changed files in src/item

pandaaaBE committed:
reupload changed files from src/data/bedrock/block/convert

pandaaaBE committed:
add NetherSprouts shears logic, cs in fungus and netherrack

pandaaaBE committed:
cs in NetherSprouts

pandaaaBE committed:
cs

pandaaaBE committed:
ups?

pandaaaBE committed:
stringtoitemparser, removed irrelevant change

pandaaaBE committed:
fix that blocks can not be spawned on non nylium blocks

pandaaaBE committed:
cs

ShockedPlot7560 commented on src/block/Fungus.php:
```diff
@@ -0,0 +1,53 @@
+	public function place(BlockTransaction $tx, Item $item, Block $blockReplace, Block $blockClicked, int $face, Vector3 $clickVector, ?Player $player = null) : bool{
+		$down = $this->getSide(Facing::DOWN);
+		$downId = $down->getTypeId();
+		if((!$down->isTransparent()) || $downId === BlockTypeIds::MYCELIUM || $downId === BlockTypeIds::PODZOL || $downId === BlockTypeIds::CRIMSON_NYLIUM || $downId === BlockTypeIds::WARPED_NYLIUM || $downId === BlockTypeIds::FLOWER_POT){
```
Unnecessary parentheses
```suggestion
		if(!$down->isTransparent() || $downId === BlockTypeIds::MYCELIUM || $downId === BlockTypeIds::PODZOL || $downId === BlockTypeIds::CRIMSON_NYLIUM || $downId === BlockTypeIds::WARPED_NYLIUM || $downId === BlockTypeIds::FLOWER_POT){
```

ShockedPlot7560 replied:
Would not be better to create a tag for that ?

pandaaaBE replied:
You mean something like in NetherRoots?

ShockedPlot7560 replied:
Like mud or dirt to detect which block can make an exception for a "mushroom" like block

pandaaaBE replied:
Okay when I am home, I will look at this

pandaaaBE replied:
@ShockedPlot7560 

now i use the "tags", but i don't know if that's what you wanted.

ShockedPlot7560 commented on src/block/NetherSprouts.php:
```diff
@@ -0,0 +1,61 @@
+
+	public function onNearbyBlockChange() : void{
+		if($this->getSide(Facing::DOWN)->isTransparent()){
+			$this->position->world->useBreakOn($this->position);
```
Use getWorld()

ShockedPlot7560 commented on src/block/NetherSprouts.php:
```diff
@@ -0,0 +1,61 @@
+	public function place(BlockTransaction $tx, Item $item, Block $blockReplace, Block $blockClicked, int $face, Vector3 $clickVector, ?Player $player = null) : bool{
+		$down = $this->getSide(Facing::DOWN);
+		$downId = $down->getTypeId();
+		if((!$down->isTransparent()) || $downId === BlockTypeIds::MYCELIUM || $downId === BlockTypeIds::PODZOL || $downId === BlockTypeIds::CRIMSON_NYLIUM || $downId === BlockTypeIds::WARPED_NYLIUM){
```
Same here

ShockedPlot7560 commented on src/block/Netherrack.php:
```diff
@@ -23,9 +23,41 @@
+	private function transformToNylium() : bool{
+		foreach($this->getHorizontalSides() as $side){
+			if($side instanceof Nylium){
+				$this->position->world->setBlock($this->position, $side);
```
Use getWorld ?

ShockedPlot7560 commented on src/block/Nylium.php:
```diff
@@ -0,0 +1,113 @@
+		$blockAbove = $this->getSide(Facing::UP);
+		if(!$blockAbove->isTransparent() && $blockAbove->getTypeId() === BlockTypeIds::SNOW_LAYER){
+			//nylium dies
+			BlockEventHelper::spread($this, VanillaBlocks::NETHERRACK(), $this);
```
Is this really a block spread ?

pandaaaBE replied:
> Similar to [grass blocks](https://minecraft.wiki/w/Grass_block) and [mycelium](https://minecraft.wiki/w/Mycelium), nylium decays (based on [random ticks](https://minecraft.wiki/w/Random_tick)) into netherrack

Should I use BlockEventHelper::die()?

ShockedPlot7560 replied:
Die isn't the right event, its use litteral for coral etc ...
Its not really a spread or a form. I don't see any created event which can handle it.

pandaaaBE replied:
Should I just replace it without a event?

ShockedPlot7560 replied:
We need this event, but in other way, Nylium is dying, however it can be confusing

pandaaaBE replied:
Should I add a function called decay()?

ShockedPlot7560 replied:
Okay, so, if we reproduce the logics of grass or mycelium. We need to call a spread event. Sry

pandaaaBE replied:
Alright xD

ShockedPlot7560 commented on src/block/RedMushroom.php:
```diff
@@ -46,8 +46,7 @@ public function place(BlockTransaction $tx, Item $item, Block $blockReplace, Blo
 		$downId = $down->getTypeId();
-		//TODO: nylium support
-		if(($lightLevel <= 12 && !$down->isTransparent()) || $downId === BlockTypeIds::MYCELIUM || $downId === BlockTypeIds::PODZOL){
+		if(($lightLevel <= 12 && !$down->isTransparent()) || $downId === BlockTypeIds::MYCELIUM || $downId === BlockTypeIds::PODZOL || $downId === BlockTypeIds::CRIMSON_NYLIUM || $downId === BlockTypeIds::WARPED_NYLIUM){
```
Same here for the tag

pandaaaBE committed:
replaced world with getWorld()

pandaaaBE committed:
replaced world with getWorld() in nylium.php and changed to BlockEventHelper::die()

pandaaaBE committed:
changed to BlockEventHelper::spread() again lol

pandaaaBE committed:
try to implement tags (??)

pandaaaBE committed:
upsi

pandaaaBE committed:
replaced world with getWorld()

ShockedPlot7560 commented on src/block/Fungus.php:
```diff
@@ -0,0 +1,40 @@
+class Fungus extends Flowable{
+	use StaticSupportTrait;
+
+	private function canBeSupportedAt(Block $block) : bool{
+		//TODO: moss
+		$supportBlock = $block->getSide(Facing::DOWN);
+		return
+			$supportBlock->hasTypeTag(BlockTypeTags::DIRT) ||
+			$supportBlock->hasTypeTag(BlockTypeTags::MUD) ||
+			$supportBlock->hasTypeTag(BlockTypeTags::NYLIUM);
```
Missing soul soil

ShockedPlot7560 commented on src/block/VanillaBlocks.php:
```diff
@@ -960,13 +965,23 @@ public function getBreakTime(Item $item) : float{
 		self::register("netherrack", new Netherrack(new BID(Ids::NETHERRACK), "Netherrack", new Info(BreakInfo::pickaxe(0.4, ToolTier::WOOD))));
 		self::register("note_block", new Note(new BID(Ids::NOTE_BLOCK, TileNote::class), "Note Block", new Info(BreakInfo::axe(0.8))));
-		self::register("obsidian", new Opaque(new BID(Ids::OBSIDIAN), "Obsidian", new Info(BreakInfo::pickaxe(35.0 /* 50 in PC */,  ToolTier::DIAMOND, 6000.0))));
+		self::register("obsidian", new Opaque(new BID(Ids::OBSIDIAN), "Obsidian", new Info(BreakInfo::pickaxe(35.0 /* 50 in PC */, ToolTier::DIAMOND, 6000.0))));
```
Unrelated change

ShockedPlot7560 commented on src/block/VanillaBlocks.php:
```diff
@@ -1317,8 +1332,7 @@ private static function registerElements() : void{
 
-		$register = fn(string $name, int $id, string $displayName, string $symbol, int $atomicWeight, int $group) =>
-			self::register("element_$name", new Element(new BID($id), $displayName, $instaBreak, $symbol, $atomicWeight, $group));
+		$register = fn(string $name, int $id, string $displayName, string $symbol, int $atomicWeight, int $group) => self::register("element_$name", new Element(new BID($id), $displayName, $instaBreak, $symbol, $atomicWeight, $group));
```
Unrelated change

ShockedPlot7560 commented on src/block/VanillaBlocks.php:
```diff
@@ -1545,7 +1559,7 @@ public function getLightLevel() : int{ return 15; }
 		self::register("warped_wart_block", new Opaque(new BID(Ids::WARPED_WART_BLOCK), "Warped Wart Block", new Info(new BreakInfo(1.0, ToolType::HOE))));
 		self::register("crying_obsidian", new class(new BID(Ids::CRYING_OBSIDIAN), "Crying Obsidian", new Info(BreakInfo::pickaxe(35.0 /* 50 in Java */, ToolTier::DIAMOND, 6000.0))) extends Opaque{
-			public function getLightLevel() : int{ return 10;}
+			public function getLightLevel() : int{ return 10; }
```
Unrelated change

ShockedPlot7560 commented on src/block/NetherSprouts.php:
```diff
@@ -0,0 +1,40 @@
+class NetherSprouts extends Flowable{
+	use StaticSupportTrait;
+
+	private function canBeSupportedAt(Block $block) : bool{
+		//TODO: moss
+		$supportBlock = $block->getSide(Facing::DOWN);
+		return
+			$supportBlock->hasTypeTag(BlockTypeTags::DIRT) ||
+			$supportBlock->hasTypeTag(BlockTypeTags::MUD) ||
+			$supportBlock->hasTypeTag(BlockTypeTags::NYLIUM);
+	}
```
Missing soul soil

pandaaaBE replied:
Couldn´t NetherSprouts extends to NetherRoots so I can delete this class?

pandaaaBE replied:
You already opened a pr to add soul soil so this would not be relevant anymore

pandaaaBE replied:
nvm fungus could extend to netherroots but sprouts can not be placed in flower pods

ShockedPlot7560 commented on src/block/Nylium.php:
```diff
@@ -0,0 +1,113 @@
+			];
+		}
+
+		$random = new Random(mt_rand());
```
No need to create a random seed

pandaaaBE replied:
to be honest, the code was taken from grass in exactly the same way...

ShockedPlot7560 replied:
Is there a reason why it was done like this ?

ShockedPlot7560 commented on src/block/RedMushroom.php:
Apply only the minimal changes, there is many unrelated changes

pandaaaBE replied:
yeah I reformated the code in vanillablocks but why shouldn't Mushrooms use this Trait too?

ShockedPlot7560 replied:
I dont know, but thats an issue for another PR, not for the moment

pandaaaBE replied:
I could rename this pr to "Implement some missing nether blocks and edit RedMushroom to use the Trait"

ShockedPlot7560 replied:
No, a PR need to done one thing by one thing

ShockedPlot7560 commented on src/block/Nylium.php:
```diff
@@ -0,0 +1,113 @@
+		/** @var Block[] $arr */
+		$arr = [];
+
+		if($this->getTypeId() === BlockTypeIds::CRIMSON_NYLIUM){
+			$arr = [
+				VanillaBlocks::CRIMSON_FUNGUS(),
+				$roots = VanillaBlocks::CRIMSON_ROOTS(),
+				$roots,
+				$roots,
+				$roots
+			];
+		}
+
+		if($this->getTypeId() === BlockTypeIds::WARPED_NYLIUM){
+			$arr = [
+				VanillaBlocks::WARPED_FUNGUS(),
+				VanillaBlocks::NETHER_SPROUTS(),
+				$roots = VanillaBlocks::WARPED_ROOTS(),
+				$roots,
+				$roots
+			];
+		}
```
The blocks which will be placed is link to the actual biome, not really the block with which you interact

pandaaaBE replied:
What do you think? I just tried it again in Minecraft and I think the behavior is identical.

ShockedPlot7560 replied:
That's what i see in the wiki. But maybe its different. Wiki says that the block placed when growing is linked to the current biome of the nylium block and not there type ? Maybe the current impl is enough

pandaaaBE committed:
Update VanillaBlocks.php

pandaaaBE committed:
Update VanillaBlocks.php

pandaaaBE committed:
removed unreleated changes in redmushroom.php

pandaaaBE committed:
add soul soil and removed the seed

pandaaaBE committed:
add soul soil to nethersprouts

pandaaaBE committed:
Delete src/NetherSprouts.php

pandaaaBE committed:
Add files via upload

pandaaaBE committed:
cs

ShockedPlot7560 commented on src/block/Fungus.php:
```diff
@@ -0,0 +1,41 @@
+		return
+			$supportBlock->hasTypeTag(BlockTypeTags::DIRT) ||
+			$supportBlock->hasTypeTag(BlockTypeTags::MUD) ||
+			$supportBlock->getTypeId() === BlockTypeIds::SOUL_SOIL ||
+			$supportBlock->hasTypeTag(BlockTypeTags::NYLIUM);
```
Could we invert the two lign ? So we have all tags before the specific block check.

ShockedPlot7560 commented on src/block/NetherRoots.php:
```diff
@@ -30,10 +30,12 @@ final class NetherRoots extends Flowable{
 			$supportBlock->hasTypeTag(BlockTypeTags::DIRT) ||
-			$supportBlock->hasTypeTag(BlockTypeTags::MUD);
+			$supportBlock->hasTypeTag(BlockTypeTags::MUD) ||
+			$supportBlock->getTypeId() === BlockTypeIds::SOUL_SOIL ||
```
Soul soil is an unrelated change for me.

ShockedPlot7560 commented on src/block/NetherSprouts.php:
```diff
@@ -0,0 +1,41 @@
+		return
+			$supportBlock->hasTypeTag(BlockTypeTags::DIRT) ||
+			$supportBlock->hasTypeTag(BlockTypeTags::MUD) ||
+			$supportBlock->getTypeId() === BlockTypeIds::SOUL_SOIL ||
+			$supportBlock->hasTypeTag(BlockTypeTags::NYLIUM);
```
Same here for the order

ShockedPlot7560 commented on src/block/Netherrack.php:
```diff
@@ -23,9 +23,41 @@
+	private function transformToNylium() : bool{
+		foreach($this->getHorizontalSides() as $side){
+			if($side instanceof Nylium){
+				$this->position->getWorld()->setBlock($this->position, $side);
```
Why did you use `$side` here, couldn't be better to use a fresh instance ?

pandaaaBE replied:
Don´t know what you mean

pandaaaBE replied:
how should I do it instead? anyways the block state never changes

pandaaaBE replied:
```
> private function transformToNylium() : bool{
> 		foreach($this->getHorizontalSides() as $side){
> 			if($side->hasTypeTag(BlockTypeTags::NYLIUM)){
> 				$block = null;
> 				if($side->getTypeId() === BlockTypeIds::CRIMSON_NYLIUM){
> 					$block = VanillaBlocks::CRIMSON_NYLIUM();
> 				}elseif($side->getTypeId() === BlockTypeIds::WARPED_NYLIUM){
> 					$block = VanillaBlocks::WARPED_NYLIUM();
> 				}
> 				
> 				$this->position->getWorld()->setBlock($this->position, $block);
> 				
> 				return true;
> 			}
> 		}
> 
> 		return false;
> 	}
```

	
Do you like this code more?

ShockedPlot7560 replied:
My fault, the original code looks ok

ShockedPlot7560 commented on src/block/Nylium.php:
```diff
@@ -0,0 +1,112 @@
+	}
+
+	public function onInteract(Item $item, int $face, Vector3 $clickVector, ?Player $player = null, array &$returnedItems = []) : bool{
+		if($this->getSide(Facing::UP)->getTypeId() !== BlockTypeIds::AIR){
```
You can't fertilize if there is already a fungus or smth like on top of a Nylium block?

pandaaaBE replied:
No? How could that work, by the way, it's still exactly the same code as Grass and they have the same behavior...

pandaaaBE committed:
removed soil sand and changed the order

pandaaaBE committed:
ups forgot to upload fungus

ShockedPlot7560 reviewed (commented):
Some PHPstan issues need to be adressed

pandaaaBE committed:
reset actual-problems.neon

pandaaaBE committed:
last try

pandaaaBE committed:
i hate my life

ShockedPlot7560 commented on tests/phpstan/configs/actual-problems.neon:
Why there is so many strange errors that have been added ?

pandaaaBE replied:
I needed to export `PATH=$PWD/bin/php7/bin:$PATH` and then run the phpstan command again

(thx to dylan)

pandaaaBE committed:
regenerated actual-problems.neon

pandaaaBE commented:
the only problem left is that the re-generated actual-problems.neon removes following thing:

```
		-
			message: "#^Parameter \\#1 \\$logFile of class pocketmine\\\\utils\\\\MainLogger constructor expects string, string\\|false given\\.$#"
			count: 1
			path: ../../phpunit/scheduler/AsyncPoolTest.php
```

Should I just copy & paste it in the file? lol

pandaaaBE commented:
Is this pr finished or is something left? I ask because of the Label

ShockedPlot7560 commented on src/block/VanillaBlocks.php:
```diff
@@ -960,6 +965,16 @@ public function getBreakTime(Item $item) : float{
+		self::register("warped_nylium", new Nylium(new BID(Ids::WARPED_NYLIUM), "Warped Nylium", $nyliumBreakInfo));
+
+		$fungusBreakInfo = new Info(BreakInfo::instant());
+		self::register("crimson_fungus", new Fungus(new BID(Ids::CRIMSON_FUNGUS), "Crimson Fungus", $fungusBreakInfo));
+		self::register("warped_fungus", new Fungus(new BID(Ids::WARPED_FUNGUS), "Warped Fungus", $fungusBreakInfo));
```
Need the `POTTABLE_PLANTS` tag. Fungus can be placed in flower pot

ShockedPlot7560 commented on src/block/Nylium.php:
```diff
@@ -0,0 +1,112 @@
+			];
+		}
+
+		$random = new Random();
```
Why using a random class instead of just `mt_rand` ?

ShockedPlot7560 commented on src/block/Nylium.php:
```diff
@@ -0,0 +1,112 @@
+		for($c = 0; $c < $count; ++$c){
+			$x = $random->nextRange($this->position->x - $radius, $this->position->x + $radius);
+			$z = $random->nextRange($this->position->z - $radius, $this->position->z + $radius);
+			if($this->position->getWorld()->getBlockAt($x, $this->position->y + 1, $z)->getTypeId() === BlockTypeIds::AIR && $this->position->getWorld()->getBlockAt($x, $this->position->y, $z) instanceof Nylium){
+				$this->position->getWorld()->setBlockAt($x, $this->position->y + 1, $z, $arr[$random->nextRange(0, $arrC)]);
+			}
```
It may be possible to reduce the entire line by using the same world variable.

ShockedPlot7560 commented on src/block/VanillaBlocks.php:
```diff
@@ -960,6 +965,16 @@ public function getBreakTime(Item $item) : float{
+		self::register("crimson_fungus", new Fungus(new BID(Ids::CRIMSON_FUNGUS), "Crimson Fungus", $fungusBreakInfo));
+		self::register("warped_fungus", new Fungus(new BID(Ids::WARPED_FUNGUS), "Warped Fungus", $fungusBreakInfo));
+
+		self::register("nether_sprouts", new NetherSprouts(new BID(Ids::NETHER_SPROUTS), "Nether Sprouts", $fungusBreakInfo));
```
So, its not really the `fungusBreakInfo`?

pandaaaBE committed:
removed random class, used $world var and some changes in vanillablocks

pandaaaBE committed:
cs

pandaaaBE commented:
Should be ready for a final review

dktapps committed:
BlockStateData: add a note about CURRENT_VERSION

this is not the same as current game version, as the revision is determined by some Mojang internal factors.

dependabot[bot] committed:
Bump shivammathur/setup-php from 2.30.4 to 2.31.1

Bumps [shivammathur/setup-php](https://github.com/shivammathur/setup-php) from 2.30.4 to 2.31.1.
- [Release notes](https://github.com/shivammathur/setup-php/releases)
- [Commits](https://github.com/shivammathur/setup-php/compare/2.30.4...2.31.1)

---
updated-dependencies:
- dependency-name: shivammathur/setup-php
  dependency-type: direct:production
  update-type: version-update:semver-minor
...

Signed-off-by: dependabot[bot] <support@github.com>

dktapps committed:
dependabot: group patch updates into a single PR

dktapps committed:
dependabot: group PHPStan dependency updates always

these typically need to be updated together anyway

dependabot[bot] committed:
Bump the phpstan group with 2 updates (#6401)

dependabot[bot] committed:
Bump docker/build-push-action from 5.3.0 to 6.3.0 (#6387)

pandaaaBE commented:
Any idea how to resolve this conflicts?

ShockedPlot7560 commented:
Resolve and regenerate the block_factory_consistency

Gaprix committed:
BedrockWorldData: Update version constants to 1.21.2 (#6399)

Co-authored-by: Dylan T <14214667+dktapps@users.noreply.github.com>
